google-cloud-core/src/main/java/com/google/cloud/batchingexperimental/RequestAccumulator.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.batchingexperimental;
+
+import com.google.api.core.InternalApi;
+import com.google.api.core.SettableApiFuture;
+import com.google.common.base.Preconditions;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Accumulates requests, used to batch many requests into one RPC call.
+ *
+ * <p>Experimental and only meant to be used by other packages in google-cloud-java.
+ *
+ * <p>Methods are not thread-safe.
+ *
+ * <p>Sample usage:
+ *
+ * <pre>{@code
+ * synchronized(accumulator) {
+ *   accumulator.add(item, future);
+ *   while (accumulator.hasBatch()) {
+ *     // If items and futures must be retained, they must be copied.
+ *     List<E> items = new ArrayList<>(accumulator.batch());
+ *     List<SettableApiFuture<R>> futures = new ArrayList<>(accumulator.futureBatch());
+ *
+ *     // ...
+ *
+ *     accumulator.next();
+ *   }
+ * }
+ * }</pre>
+ */
+@InternalApi
+public class RequestAccumulator<E, R> {
+  private final ArrayList<E> requests = new ArrayList<>();
+  private final ArrayList<SettableApiFuture<R>> futures = new ArrayList<>();
```

Comment:
Might be nicer to have a value class encapsulate a request/future pair instead of managing parallel lists

Reply:
Won't this move the complexity elsewhere though? If we return a `List<Pair>` then the user will have to split the pair themselves when they want to send RPCs.

Perhaps you meant we keep the current surface and create `List<E>` from `List<Pair>` in `batch()`. I'm not strongly against this idea, though it seems wasteful to create new lists and pair objects often. I'm probably prematurely optimizing. In this case, I think it's somewhat OK. The class is small enough that it isn't hard to comb through, and it's easy to change if need arises.